File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Complex/Table/Table.cs
﻿using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using JDI_Commons;
using JDI_Web.Attributes.Objects;
using JDI_Web.Selenium.Elements.APIInteract;
using JDI_Web.Selenium.Elements.Base;
using JDI_Web.Selenium.Elements.Common;
using JDI_Web.Selenium.Elements.Complex.Table.Interfaces;
using OpenQA.Selenium;
using RestSharp.Extensions;
using static Epam.JDI.Core.Settings.JDISettings;
using static JDI_Web.Selenium.Elements.Complex.Table.Column;
using static JDI_Web.Selenium.Elements.Complex.Table.Row;

namespace JDI_Web.Selenium.Elements.Complex.Table
{
    public class Table : Text, ITable
    {
        private IList<ICell> _allCells = new List<ICell>();
        private readonly Columns _columns = new Columns();
        private readonly Rows _rows = new Rows();
        private IList<string> _footer;

        private IList<ICell> AllCells
        {
            get
            {
                var result = new List<ICell>();
                var rows = Rows.Get();
                Columns.Headers.ForEach(columnName => Rows.Headers
                    .ForEach(rowName => result.Add(rows[rowName][columnName])));

                if (Cache)
                    AllCells = result;
                return result;
            }
            set { _allCells = value; }
        }

        public Columns Columns
        {
            get { return _columns; }
            set { _columns.Update(value); }
        }

        public Rows Rows
        {
            get { return _rows; }
            set { _rows.Update(value); }
        }

        public bool Cache { set; get; }

        public By RootBy { set { WebAvatar.ByLocator = value;  } }
        protected By CellLocatorBy { set; get; }
        public By ColumnBy { set { Columns.LineTemplate = value; } }
        public By RowBy { set { Rows.LineTemplate = value; } }
        public By FooterBy { get; set; } = By.XPath(".//tfoot/tr/th");
        public By ColumnHeaderBy { set { Columns.HeadersLocator = value; } }
        public By RowHeaderBy { set { Rows.HeadersLocator = value; } }

        public string[] ColumnHeaders
        {
            set
            {
                Columns.AddHeaders(value);
                Columns.HasHeader = true;
            }
        }

        public string[] RowHeaders
        {
            set
            {
                Rows.AddHeaders(value);
                Rows.HasHeader = true;
            }
        }

        public int ColumnFrom { set { Columns.StartIndex = value; } }
        public int RowFrom { set { Rows.StartIndex = value; } }

        public TableHeaderTypes HeaderType
        {
            set
            {
                switch (value)
                {
                    case TableHeaderTypes.AllHeaders:
                        Columns.HasHeader = true;
                        Rows.HasHeader = true;
                        break;
                    case TableHeaderTypes.NoHeaders:
                        Columns.HasHeader = false;
                        Rows.HasHeader = false;
                        break;
                    case TableHeaderTypes.ColumnsHeaders:
                        Columns.HasHeader = true;
                        Rows.HasHeader = false;
                        break;
                    case TableHeaderTypes.RowsHeaders:
                        Columns.HasHeader = false;
                        Rows.HasHeader = true;
                        break;
                    default:
                        throw new ArgumentOutOfRangeException(nameof(value), value, null);
                }
            }
        }

        public Table()
        {
            Columns.Table = this;
            Rows.Table = this;
        }


        public Table Copy()
        {
            return Clone();
        }

        public Table Clone()
        {
            var newTable = new Table();
            newTable.Rows = Rows.Clone(new Rows(), newTable);
            newTable.Columns = Columns.Clone(new Columns(), newTable);
            newTable.WebAvatar = new GetElementModule(newTable, Locator);
            newTable.Parent = Parent;
            return newTable;
        }

        public void SetTableSettings(TableSettings settings)
        {
            Rows.HasHeader = settings.RowHasHeaders;
            Rows.Headers = settings.RowHeaders;
            Rows.Count = settings.RowsCount;
            Columns.HasHeader = settings.ColumnHasHeaders;
            Columns.Headers = settings.ColumnHeaders;
            Columns.Count = settings.ColumnsCount;
        }

        public ITable SetUp(By root, By cell, By row, By column, By footer, int colStartIndex, int rowStartIndex)
        {
            SetAvatar(byLocator: root);
            CellLocatorBy = cell;
            Rows.LineTemplate = row;
            Columns.LineTemplate = column;
            FooterBy = footer;
            Columns.StartIndex = colStartIndex;
            Rows.StartIndex = rowStartIndex;
            return this;
        }

        public ITable UseCache()
        {
            Cache = true;
            return this;
        }

        public ITable UseCache(bool value)
        {
            Cache = value;
            return this;
        }

        public void Clean()
        {
            AllCells = new List<ICell>();
            Columns.Clean();
            Rows.Clean();
        }

        public void Clear()
        {
            Clean();
        }

        Columns ITable.Columns()
        {
            throw new NotImplementedException();
        }

        public Dictionary<string, ICell> Column(int colNum)
        {
            return Columns.GetColumn(colNum);
        }

        public Dictionary<string, ICell> Column(string colName)
        {
            return Columns.GetColumn(colName);
        }

        public Dictionary<string, ICell> Column(string value, Row row)
        {
            var columnCell = Cell(value, row);
            return columnCell != null ? Columns.GetColumn(columnCell.ColumnNum) : null;
        }

        public Dictionary<string, ICell> Row(string value, Column column)
        {
            var rowCell = Cell(value, column);
            return rowCell != null ? Rows.GetRow(rowCell.RowNum) : null;
        }

        public IList<string> ColumnValue(int colNum)
        {
            return Columns.GetColumnValue(colNum);
        }

        public IList<string> ColumnValue(string colName)
        {
            return Columns.GetColumnValue(colName);
        }

        private Dictionary<string, ICell> Column(Column column)
        {
            return column.Get(Column, Column);
        }

        public Dictionary<string, ICell> Row(int rowNum)
        {
            return Rows.GetRow(rowNum);
        }

        public Dictionary<string, ICell> Row(string rowName)
        {
            return Rows.GetRow(rowName);
        }

        public IList<string> RowValue(int rowNum)
        {
            return Rows.GetRowValue(rowNum);
        }

        public IList<string> RowValue(string rowName)
        {
            return Rows.GetRowValue(rowName);
        }

        private Dictionary<string, ICell> Row(Row row)
        {
            return row.Get(Row, Row);
        }

        private IList<string> GetAllEnumNames(Type headers)
        {
            return typeof(Enum).IsAssignableFrom(headers) 
                ? Enum.GetNames(headers).ToList() 
                : headers.GetFields().ToList().Select(el => el.GetValue(null).ToString()).ToList();
        }

        public ITable SetColumnsCount(int value)
        {
            Columns.Count = value;
            return this;
        }

        public ITable SetRowsCount(int value)
        {
            Rows.Count = value;
            return this;
        }

        protected IList<string> GetFooterAction()
        {
            return WebElement.FindElements(FooterBy).Select(e => e.Text).ToList();
        }


        public IList<string> Footer
        {
            get { return new List<string>(_footer); }
            set { _footer = new List<string>(value); }
        }

        public Dictionary<string, SelectableElement> Header()
        {
            return Columns.Header();
        }

        public SelectableElement Header(string name)
        {
            return Columns.Header(name);
        }

        public IList<string> Headers => Columns.Headers;

        IList<string> ITable.Footer()
        {
            throw new NotImplementedException();
        }

        public IList<string> FooterInstance() // TODO method name conflict
        {
            if (Footer != null)
                return Footer;
            //_footer = invoker.doJActionResult("Get Footer", this::getFooterAction); TODO
            if (Footer == null || Footer.Count == 0)
                return new List<string>();
            Columns.Count = Footer.Count;
            return Footer;
        }

        public ICell Cell(Column column, Row row)
        {
            int colIndex = column.Get(GetColumnIndex, num => num + Columns.StartIndex - 1);
            int rowIndex = (int) row.Get(GetRowIndex, num => num + Rows.StartIndex - 1);
            return AddCell(colIndex, rowIndex, column.Get(name => Columns.Headers.IndexOf(name) + 1, num => num), row.Get(name => Rows.Headers.IndexOf(name) + 1, num => num), column.Get(name => name, num => ""), row.Get(name => name, num => ""));
        }

        public ICell Cell(string columnName, string rowName)
        {
            return Cell(column(columnName), row(rowName));
        }

        public ICell Cell(int columnIndex, int rowIndex)
        {
            return Cell(column(columnIndex), row(rowIndex));
        }

        public ICell Cell(IWebElement webElement, Column column, Row row)
        {
            return AddCell(webElement, column.Get(name => Columns.Headers.IndexOf(name) + 1, num => num), row.Get(name => Rows.Headers.IndexOf(name) + 1, num => num), column.Get(name => name, num => ""), row.Get(name => name, num => ""));
        }

        private IList<ICell> Matches(Collection<ICell> list, string regex)
        {
            return new List<ICell>(list.Where(cell => cell.Value.Matches(regex)));
        }

        public IList<ICell> Cells(string value)
        {
            return new List<ICell>(GetCells().Where(cell => cell.Value.Equals(value)));
        }


        public ICell Cell(string value)
        {
            return Rows.Get().Select(row => row.Value.FirstOrDefault(pair => pair.Value.GetText.Equals(value)).Value).FirstOrDefault(result => result != null);
        }

        public IList<ICell> GetCells()
        {
            var rows = Rows.Get();
            var result = (from columnName in Columns.Headers from rowName in Rows.Headers select rows[rowName][columnName]).ToList();
            if (Cache)
                AllCells = result;
            return result;
        }

        public Dictionary<string, Dictionary<string, ICell>> GetRows(params string[] colNameValues)
            // TODO method name conflict
        {
            var result = new Dictionary<string, Dictionary<string, ICell>>();
            foreach (var row in Rows.Get())
            {
                var matches = true;
                foreach (var colNameValue in colNameValues)
                {
                    if (!colNameValue.Matches("[^=]+=[^=]*"))
                        throw Exception($"Wrong searchCriteria for Cells: {colNameValue}");
                    var splitted = colNameValue.Split(Convert.ToChar("="));
                    var colName = splitted[0];
                    var colValue = splitted[1];
                    var cell = row.Value[colName];
                    if (cell == null || !cell.Value.Equals(colValue))
                    {
                        matches = false;
                        break;
                    }
                }
                if (matches) result.Add(row.Key, row.Value);
            }
            return result;
        }

        public Dictionary<string, Dictionary<string, ICell>> GetColumns(params string[] rowNameValues)
            // TODO method name conflict
        {
            var result = new Dictionary<string, Dictionary<string, ICell>>();
            foreach (var column in Columns.Get())
            {
                var matches = true;
                foreach (var rowNameValue in rowNameValues)
                {
                    if (!rowNameValue.Matches("[^=]+=[^=]*"))
                        throw Exception($"Wrong searchCritaria for Cells: {rowNameValue}");
                    var splitted = rowNameValue.Split(Convert.ToChar("="));
                    var rowName = splitted[0];
                    var rowValue = splitted[1];
                    var cell = column.Value[rowName];
                    if (cell == null || !cell.Value.Equals(rowValue))
                    {
                        matches = false;
                        break;
                    }
                }
                if (matches) result.Add(column.Key, column.Value);
            }
            return result;
        }


        public bool WaitValue(string value, Row row)
        {
            return Timer.Wait(() => Column(value, row) != null);
        }

        public bool WaitValue(string value, Column column)
        {
            return Timer.Wait(() => Row(value, column) != null);
        }

        public bool Empty
        {
            get
            {
                WebDriver.Manage().Timeouts().ImplicitlyWait(TimeSpan.Zero);
                var rowsCount = Rows.GetCount(true);
                WebDriver.Manage().Timeouts().ImplicitlyWait(TimeSpan.FromSeconds(Timeouts.CurrentTimeoutSec));
                return rowsCount == 0;
            }
        }

        public bool WaitHaveRows()
        {
            return WaitRows(1);
        }

        public bool WaitRows(int count)
        {
            return Timer.Wait(() => Rows.Count >= count);
        }

        public ICell Cell(string value, Column column)
        {
            var colIndex = column.Get(name => Columns.Headers.IndexOf(name) + 1, num => num);
            return Columns.GetColumn(colIndex).FirstOrDefault(pair => pair.Value.Value.Equals(value)).Value;
        }

        public ICell Cell(string value, Row row)
        {
            var rowNum = row.HasName ? Rows.Headers.IndexOf(row.Name) + 1 : row.Num;
            return Rows.GetRow(rowNum).FirstOrDefault(pair => pair.Value.Value.Equals(value)).Value;
        }

        public IList<ICell> CellsMatch(string regex)
        {
            return Matches(new Collection<ICell>(GetCells()), regex);
        }

        public ICell CellMatch(string regex)
        {
            return Rows.Get().Select(row => row.Value.FirstOrDefault(pair => pair.Value.GetText.Matches(regex)).Value).FirstOrDefault(result => result != null);
        }

        public IList<ICell> CellsMatch(string regex, Column column)
        {
            var columnLine = Column(column);
            return new List<ICell>(((Dictionary<string, ICell>) columnLine.Where(v => v.Key.Matches(regex))).Values);
        }

        public IList<ICell> CellsMatch(string regex, Row row)
        {
            var columnLine = Row(row);
            return new List<ICell>(((Dictionary<string, ICell>) columnLine.Where(v => v.Key.Matches(regex))).Values);
        }

        private int GetColumnIndex(string name)
        {
            int nameIndex;
            var headers = Columns.Headers;
            if (headers != null && headers.Contains(name))
                nameIndex = headers.IndexOf(name);
            else
                throw Exception("Can't Get Column: '" + name + "'. " + ((headers == null) ? "ColumnHeaders is Null" : "Available ColumnHeaders: " + headers.Print(", ", "'{0}'") + ")"));
            return nameIndex + Columns.StartIndex;
        }

        private object GetRowIndex(string name)
        {
            int nameIndex;
            var headers = Rows.Headers;
            if (headers != null && headers.Contains(name))
                nameIndex = headers.IndexOf(name);
            else
                throw Exception($"Can't Get Row: {name}. Available RowHeaders: {Headers.Print(", ", "'{0}'")}");
            return nameIndex + Rows.StartIndex;
        }

        protected new string GetTextAction()
        {
            return "||X||" + Columns.Headers.Print("|") + "||\n" + Rows.Headers.Select(rowName => "||" + rowName + "||" + GetCells().Where(cell => cell.RowName.Equals(rowName)).Select(cell => cell.Value).Print("|") + "||").Print("\n");
        }

        private ICell AddCell(int colIndex, int rowIndex, int colNum, int rowNum, string colName, string rowName)
        {
            var cell = (Cell) AllCells.FirstOrDefault(c => c.ColumnNum == colNum && c.RowNum == rowNum);
            if (cell != null)
                return cell.UpdateData(colName, rowName);
            cell = new Cell(colNum, rowNum, colName, rowName, CellLocatorBy, this, colIndex, rowIndex);
            cell.SetAvatar((GetElementModule) cell.Get().Avatar);

            if (Cache)
                AllCells.Add(cell);
            return cell;
        }

        private ICell AddCell(IWebElement webElement, int colNum, int rowNum, string colName, string rowName)
        {
            var cell = (Cell) _allCells.FirstOrDefault(c => c.ColumnNum == colNum && c.RowNum == rowNum);
            if (cell != null)
            {
                cell.WebElement = webElement;
                return cell.UpdateData(colName, rowName);
            }
            cell = new Cell(colNum, rowNum, colName, rowName, CellLocatorBy, this, -1, webElement: webElement);

            if (Cache)
                AllCells.Add(cell);
            return cell;
        }
    }
}

File: Java/JDI/jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winnium/elements/complex/table/Table.java
package com.epam.jdi.uitests.win.winnium.elements.complex.table;

import com.epam.commons.map.MapArray;
import com.epam.commons.pairs.Pair;
import com.epam.jdi.uitests.core.interfaces.base.ISelect;
import com.epam.jdi.uitests.core.interfaces.complex.tables.interfaces.*;
import com.epam.jdi.uitests.core.settings.JDISettings;
import com.epam.jdi.uitests.win.winnium.elements.ElementsUtils;
import com.epam.jdi.uitests.win.winnium.elements.apiInteract.GetElementModule;
import com.epam.jdi.uitests.win.winnium.elements.base.Element;
import com.epam.jdi.uitests.win.winnium.elements.pageobjects.annotations.objects.TableHeaderTypes;
import org.openqa.selenium.By;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.epam.commons.EnumUtils.getAllEnumNames;
import static com.epam.commons.PrintUtils.print;
import static com.epam.commons.Timer.getByCondition;
import static com.epam.commons.Timer.waitCondition;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;
import static java.lang.String.format;

public class Table extends Element implements ITable {
    private boolean useCache = true;
    private String cellXpath;

    protected Columns columns;
    protected Rows rows;

    public Table() {
        columns = new Columns(this);
        rows = new Rows(this);
    }

    public boolean isUseCache() {
        return useCache;
    }

    public void setUp(By root, String columnHeadersInTableXpath, String rowsInTableXpath, String headerInRowXpath,
                      String columnsInRowXpath, TableHeaderTypes tableHeaderTypes, int rowStartIndex,
                      int colStartIndex, boolean useCache) {
        cellXpath = "." + rowsInTableXpath + "[%d]" + columnsInRowXpath + "[%d]";

        if (root != null)
            setAvatar(new GetElementModule(root, this));

        this.useCache = useCache;

        rows.setStartIndex(rowStartIndex);
        rows.setLineTemplate("." + rowsInTableXpath + "[%d]" + columnsInRowXpath);
        if (!"".equals(rowsInTableXpath) || !"".equals(headerInRowXpath))
            rows.setHeadersXpathStr("." + rowsInTableXpath + headerInRowXpath);
        columns.setStartIndex(colStartIndex);
        columns.setLineTemplate("." + rowsInTableXpath + columnsInRowXpath + "[%d]");
        if (!"".equals(columnHeadersInTableXpath))
            columns.setHeadersXpathStr("." + columnHeadersInTableXpath);

        switch (tableHeaderTypes) {
            case ALL_HEADERS:
                hasAllHeaders();
                break;
            case NO_HEADERS:
                hasNoHeaders();
                break;
            case COLUMNS_HEADERS:
                hasOnlyColumnHeaders();
                break;
            case ROWS_HEADERS:
                hasOnlyRowHeaders();
                break;
        }
    }

    private int getColumnIndex(String name) {
        List<String> headers = columns().headers();
        if (headers != null) {
            int nameIndex = headers.indexOf(name);
            if (nameIndex != -1)
                return nameIndex + columns().getStartIndex();
        }

        throw exception("Can't Get Column: '" + name + "'. " + ((headers == null)
                ? "ColumnHeaders is Null"
                : ("Available ColumnHeaders: " + print(headers, ", ", "'{0}'") + ")")));
    }

    private int getRowIndex(String name) {
        int nameIndex;
        List<String> headers = rows().headers();
        if (headers != null && headers.contains(name))
            nameIndex = headers.indexOf(name);
        else
            throw exception("Can't Get Row: '%s'. Available RowHeaders: (%s)", name, print(headers, ", ", "'{0}'"));
        return nameIndex + rows().getStartIndex();
    }

    @Override
    public ICell cell(Column column, Row row) {
        int colIndex = column.get(this::getColumnIndex, num -> num + columns().getStartIndex() - 1);
        int rowIndex = row.get(this::getRowIndex, num -> num + rows().getStartIndex() - 1);

        return addCell(colIndex, rowIndex,
                column.get(name -> columns().headers().indexOf(name) + 1, num -> num),
                row.get(name -> rows().headers().indexOf(name) + 1, num -> num),
                column.get(name -> name, num -> ""),
                row.get(name -> name, num -> ""));
    }

    private ICell addCell(int colIndex, int rowIndex, int colNum, int rowNum, String colName, String rowName) {
        return new Cell(colIndex, rowIndex, colNum, rowNum, colName, rowName, cellXpath, this);
    }

    @Override
    public List<ICell> cells(String value) {
        return getCells().stream().filter(cell -> cell.getValue().equals(value)).collect(Collectors.toList());
    }

    @Override
    public ICell cellContains(String value, Row row) {
        int rowNum = row.hasName() ? rows().headers().indexOf(row.getName()) + 1 : row.getNum();

        return rows().getRow(rowNum).first((name, cell) -> cell.getValue().contains(value));
    }

    @Override
    public ICell cellMatch(String regex, Row row) {
        MapArray<String, ICell> rowLine = row(row);
        Optional<ICell> cellOpt = rowLine.values().stream().filter(cell -> cell.getValue().matches(regex)).findAny();

        return cellOpt.isPresent() ? cellOpt.get() : null;
    }

    private MapArray<String, ICell> row(Row row) {
        return row.get(this::row, this::row);
    }

    @Override
    public ICell cellContains(String value, Column column) {
        int colIndex = column.get(
                name -> columns().headers().indexOf(name) + 1,
                num -> num);
        return columns().getColumn(colIndex).first((name, cell) -> cell.getValue().contains(value));
    }

    @Override
    public ICell cellMatch(String regex, Column column) {
        MapArray<String, ICell> columnLine = column(column);
        Optional<ICell> cellOpt = columnLine.values().stream().filter(cell -> cell.getValue().matches(regex)).findAny();

        return cellOpt.isPresent() ? cellOpt.get() : null;
    }

    private MapArray<String, ICell> column(Column column) {
        return column.get(this::column, this::column);
    }

    @Override
    public List<ICell> cellsContains(String value) {
        return getCells().stream().filter(cell -> cell.getValue().equals(value)).collect(Collectors.toList());
    }

    @Override
    public List<ICell> cellsMatch(String regex) {
        return getCells().stream().filter(cell -> cell.getValue().matches(regex)).collect(Collectors.toList());
    }

    @Override
    public ICell cell(String value) {
        Optional<ICell> cellOpt = getCells().stream().filter(cell -> cell.getValue().equals(value)).findAny();
        return cellOpt.isPresent() ? cellOpt.get() : null;
    }

    @Override
    public ICell cellMatch(String regex) {
        Optional<ICell> cellOpt = getCells().stream().filter(cell -> cell.getValue().matches(regex)).findAny();
        return cellOpt.isPresent() ? cellOpt.get() : null;
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> rows(String... colNameValues) {
        List<TableFilter> filters = new ArrayList<>();
        for (String colNameValue : colNameValues)
            filters.add(new TableFilter(colNameValue));

        boolean matches = false;
        MapArray<String, MapArray<String, ICell>> result = new MapArray<>();
        for (Pair<String, MapArray<String, ICell>> row : rows().get()) {
            for (TableFilter filter : filters) {
                ICell cell = row.value.get(filter.name);
                if (cell == null)
                    throw exception(format("Search rows for '%s' failed. Can't get cell for column named %s",
                            print(colNameValues), filter.name));
                switch (filter.type) {
                    case EQUAL:
                        matches = cell.getValue().equals(filter.value);
                        break;
                    case CONTAINS:
                        matches = cell.getValue().contains(filter.value);
                        break;
                    default:
                        matches = cell.getValue().matches(filter.value);
                        break;
                }
                if (!matches) break;
            }
            if (matches) result.add(row);
        }
        return result;
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> rows(String value, Column column) {
        return rows().withValue(value, column);
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> rowsContains(String value, Column column) {
        return rows().containsValue(value, column);
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> rowsMatches(String regEx, Column column) {
        return rows().matchesRegEx(regEx, column);
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> columns(String... rowNameValues) {
        MapArray<String, MapArray<String, ICell>> result = new MapArray<>();
        for (Pair<String, MapArray<String, ICell>> column : columns().get()) {
            boolean matches = true;
            for (String rowNameValue : rowNameValues) {
                if (!rowNameValue.matches("[^=]+=[^=]*"))
                    throw exception("Wrong searchCritaria for Cells: " + rowNameValue);
                String[] splitted = rowNameValue.split("=");
                String rowName = splitted[0];
                String rowValue = splitted[1];
                ICell cell = column.value.get(rowName);
                if (cell == null || !cell.getValue().equals(rowValue)) {
                    matches = false;
                    break;
                }
            }
            if (matches) result.add(column);
        }
        return result;
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> columns(String value, Row row) {
        return columns().withValue(value, row);
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> columnsContains(String value, Row row) {
        return columns().containsValue(value, row);
    }

    @Override
    public MapArray<String, MapArray<String, ICell>> columnsMatches(String regEx, Row row) {
        return columns().withValue(regEx, row);
    }

    @Override
    public boolean waitValue(String value, Row row) {
        return ElementsUtils.timer().wait(() -> column(value, row) != null);
    }

    @Override
    public boolean waitValue(String value, Column column) {
        return ElementsUtils.timer().wait(() -> row(value, column) != null);
    }

    @Override
    public boolean isEmpty() {
        return rows().count() == 0;
    }

    @Override
    public boolean waitHaveRows() {
        return waitRows(1);
    }

    @Override
    public boolean waitRows(int count) {
        return waitCondition(() -> rows().count() >= count);
    }

    @Override
    public ICell cell(String value, Row row) {
        int rowIndex = row.get(name -> rows().headers().indexOf(name) + 1, num -> num);
        return rows().getRow(rowIndex).first((name, cell) -> cell.getValue().equals(value));
    }

    @Override
    public ICell cell(String value, Column column) {
        int colIndex = column.get(name -> columns().headers().indexOf(name) + 1, num -> num);
        return columns().getColumn(colIndex).first((name, cell) -> cell.getValue().equals(value));
    }

    @Override
    public List<ICell> cellsMatch(String regex, Row row) {
        MapArray<String, ICell> rowLine = row(row);
        return rowLine.values().stream().filter(cell -> cell.getValue().matches(regex)).collect(Collectors.toList());
    }

    @Override
    public List<ICell> cellsMatch(String regex, Column column) {
        MapArray<String, ICell> columnLine = column(column);
        return columnLine.values().stream().filter(cell -> cell.getValue().matches(regex)).collect(Collectors.toList());
    }

    @Override
    public MapArray<String, ICell> row(String value, Column column) {
        ICell rowCell = cell(value, column);
        return rowCell != null ? rows().getRow(rowCell.rowNum()) : null;
    }

    @Override
    public MapArray<String, ICell> column(String value, Row row) {
        ICell columnCell = cellContains(value, row);
        return columnCell != null ? columns().getColumn(columnCell.columnNum()) : null;
    }

    @Override
    public MapArray<String, ICell> columnContains(String value, Row row) {
        ICell columnCell = cellContains(value, row);
        return columnCell != null ? columns().getColumn(columnCell.columnNum()) : null;
    }

    @Override
    public MapArray<String, ICell> columnMatch(String regEx, Row row) {
        ICell columnCell = cellMatch(regEx, row);
        return columnCell != null ? columns().getColumn(columnCell.columnNum()) : null;
    }

    @Override
    public MapArray<String, ICell> rowContains(String value, Column column) {
        ICell rowCell = cellContains(value, column);
        return rowCell != null ? rows().getRow(rowCell.rowNum()) : null;
    }

    @Override
    public MapArray<String, ICell> rowMatch(String regEx, Column column) {
        ICell rowCell = cellMatch(regEx, column);
        return rowCell != null ? rows().getRow(rowCell.rowNum()) : null;
    }

    @Override
    public IRow rows() {
        return rows;
    }

    @Override
    public MapArray<String, ICell> row(int rowNum) {
        return rows().getRow(rowNum);
    }

    @Override
    public MapArray<String, ICell> row(String rowName) {
        return rows().getRow(rowName);
    }

    @Override
    public List<String> rowValue(int rowNum) {
        return rows().getRowValue(rowNum);
    }

    @Override
    public List<String> rowValue(String rowName) {
        return rows().getRowValue(rowName);
    }

    @Override
    public IColumn columns() {
        return columns;
    }

    @Override
    public MapArray<String, ICell> column(int colNum) {
        return columns().getColumn(colNum);
    }

    @Override
    public MapArray<String, ICell> column(String colName) {
        return columns().getColumn(colName);
    }

    @Override
    public List<String> columnValue(int colNum) {
        return columns().getColumnValue(colNum);
    }

    @Override
    public List<String> columnValue(String colName) {
        return columns().getColumnValue(colName);
    }

    @Override
    public MapArray<String, ISelect> header() {
        return columns().header();
    }

    @Override
    public ISelect header(String name) {
        return columns().header(name);
    }

    @Override
    public List<String> headers() {
        return columns.headers();
    }

    @Override
    public List<String> footer() {
        throw JDISettings.exception("Not supported");
    }

    @Override
    public List<ICell> getCells() {
        List<ICell> result = new ArrayList<>();
        for (String columnName : columns().headers())
            rows().headers().forEach(rowName
                    -> result.add(cell(columnName, rowName)));

        return result;
    }

    @Override
    public void clean() {
        columns().clean();
        rows().clean();
    }

    @Override
    public void clear() {
        clean();
    }

    @Override
    public ITable useCache(boolean value) {
        useCache = value;
        return this;
    }

    @Override
    public ITable clone() {
        throw JDISettings.exception("Not supported");
    }

    @Override
    public ITable copy() {
        return clone();
    }

    public ITable hasAllHeaders() {
        ((Columns) columns()).setHasHeader(true);
        ((Rows)rows()).setHasHeader(true);
        return this;
    }

    public ITable hasNoHeaders() {
        ((Columns) columns()).setHasHeader(false);
        ((Rows)rows()).setHasHeader(false);
        return this;
    }

    public ITable hasOnlyColumnHeaders() {
        ((Columns) columns()).setHasHeader(true);
        ((Rows)rows()).setHasHeader(false);
        return this;
    }

    public ITable hasOnlyRowHeaders() {
        ((Columns) columns()).setHasHeader(false);
        ((Rows)rows()).setHasHeader(true);
        return this;
    }

    @Override
    public ITable hasColumnHeaders(List<String> value) {
        ((Columns) columns()).setHasHeader(true);
        ((Columns) columns()).headers = new ArrayList<>(value);
        return this;
    }

    @Override
    public <THeaders extends Enum> ITable hasColumnHeaders(Class<THeaders> headers) {
        return hasColumnHeaders(getAllEnumNames(headers));
    }

    @Override
    public ITable hasRowHeaders(List<String> value) {
        ((Rows)rows()).setHasHeader(true);
        ((Rows)rows()).headers = new ArrayList<>(value);
        return this;
    }

    @Override
    public <THeaders extends Enum> ITable hasRowHeaders(Class<THeaders> headers) {
        return hasRowHeaders(getAllEnumNames(headers));
    }

    @Override
    public ITable setColumnsCount(int value) {
        columns().setCount(value);
        return this;
    }

    @Override
    public ITable setRowsCount(int value) {
        rows().setCount(value);
        return this;
    }

    @Override
    public String getText() {
        return "||X||" + print(columns().headers(), "|") + "||\n" + print(
                rows().headers().stream().map(rowName -> "||" + rowName + "||" + print(rowValue(rowName), "|") + "||").
                        collect(Collectors.toList()), "\n");
    }

    @Override
    public String waitText(String text) {
        return invoker.doJActionResult(format("Wait text contains '%s'", text),
                () -> getByCondition(this::getText, t -> t.contains(text)), this.toString());
    }

    @Override
    public String waitMatchText(String regEx) {
        return invoker.doJActionResult(format("Wait text match regex '%s'", regEx),
                () -> getByCondition(this::getText, t -> t.matches(regEx)), this.toString());
    }

    @Override
    public String getValue() {
        return getText();
    }
}
